man/gpxlogger.1
'\" t
.\"     Title: gpxlogger
.\"    Author: [see the "AUTHORS" section]
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 6 December 2020
.\"    Manual: GPSD Documentation
.\"    Source: The GPSD Project
.\"  Language: English
.\"
.TH "GPXLOGGER" "1" "6 December 2020" "The GPSD Project" "GPSD Documentation"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
gpxlogger \- Tool to connect to gpsd and generate a GPX file
.SH "SYNOPSIS"
.HP \w'\fBgpxlogger\fR\ 'u
\fBgpxlogger\fR [\-?] [\-\-daemonize] [\-\-debug\ \fIdebug\-level\fR] [\-\-export\ \fIexport\-method\fR] [\-\-exports] [\-\-help] [\-\-interval\fItrack\ timeout\fR] [\-\-minmove\ \fIminmove\fR] [\-\-output\ \fIfilename\fR] [\-\-version] [\-d] [\-D\ \fIdebug\-level\fR] [\-e\ \fIexport\-method\fR] [\-f\ \fIfilename\fR] [\-h] [\-i\ \fItrack\ timeout\fR] [\-l] [\-m\ \fIminmove\fR] [\-V] [\fIserver\fR [\fI:port\fR [\fI:device\fR]]]
.SH "DESCRIPTION"
.PP
This program collects fixes from
gpsd
and logs them to standard output in GPX, an XML profile for track logging\&.
.PP
The output may be composed of multiple tracks\&. A new track is created if there\*(Aqs no fix written for an interval specified by the
\fB\-i, \-\-interval\fR
and defaulting to 5 seconds\&.
.PP
gpxlogger
can use any of the export methods that
gpsd
supports\&. For a list of these methods, use the
\fB\-l, \-\-exports\fR\&. To force the method, give the
\fB\-e, \-\-export\fR
one of the colon\-terminated method names from the
\fB\-l, \-\-exports\fR
table\&.
.SH "OPTIONS"
.PP
\fB\-?\fR, \fB\-h\fR, \fB\-\-help\fR
.RS 4
Print a summary of options and then exit\&.
.RE
.PP
\fB\-d\fR, \fB\-\-daemonize\fR
.RS 4
Run as a daemon in background\&. It requires the
\fB\-f, \-\-output\fR
option, which directs output to a specified logfile\&.
.RE
.PP
\fB\-D LVL\fR, \fB\-\-debug LVL\fR
.RS 4
Sets the debug level; it is primarily for troubleshooting\&. It enables various progress messages to standard error\&.
.RE
.PP
\fB\-e METHOD\fR, \fB\-\-export METHOD\fR
.RS 4
If D\-Bus support is available on the host, GPSD is configured to use it, and
\fB\-e dbus, \-\-export dbus\fR
is specified, this program listens to DBUS broadcasts from
gpsd
via org\&.gpsd\&.fix\&.
.sp
With
\fB\-e sockets\fR, or if sockets is the method defaulted to, you may give a server\-port\-device specification as arguments\&.
.sp
The sockets default is to all devices on the localhost, using the default GPSD port 2947\&. An optional argument to any client may specify a server to get data from\&. A colon\-separated suffix is taken as a port number\&. If there is a second colon\-separated suffix, that is taken as a specific device name to be watched\&. However, if the server specification contains square brackets, the part inside them is taken as an IPv6 address and port/device suffixes are only parsed after the trailing bracket\&. Possible cases look like this:
.PP
localhost:/dev/ttyS1
.RS 4
Look at the default port of localhost, trying both IPv4 and IPv6 and watching output from serial device 1\&.
.RE
.PP
example\&.com:2317
.RS 4
Look at port 2317 on example\&.com, trying both IPv4 and IPv6\&.
.RE
.PP
71\&.162\&.241\&.5:2317:/dev/ttyS3
.RS 4
Look at port 2317 at the specified IPv4 address, collecting data from attached serial device 3\&.
.RE
.PP
[FEDC:BA98:7654:3210:FEDC:BA98:7654:3210]:2317:/dev/ttyS5
.RS 4
Look at port 2317 at the specified IPv6 address, collecting data from attached serial device 5\&.
.RE
.sp
With
\fB\-e shm, \-\-export shm\fR
this program will listen to the local
gpsd
using shared memory\&.
.RE
.PP
\fB\-i SECONDS\fR, \fB\-\-interval SECONDS\fR
.RS 4
A new track is created if there\*(Aqs no fix written for an interval of SECONDS\&. Defaulting to 5 seconds\&.
.RE
.PP
\fB\-l\fR, \fB\-\-exports\fR
.RS 4
List all possible options for
\fB\-e, \-\-export\fR\&.
.RE
.PP
\fB\-m MINMOVE\fR, \fB\-\-minmove MINMOVE\fR
.RS 4
Sets a minimum move distance in meters (it may include a fractional decimal part)\&. Motions shorter than this will not be logged\&.
.RE
.PP
\fB\-r\fR, \fB\-\-reconnect\fR
.RS 4
Retry when GPSd loses the fix\&. Without
\fB\-r\fR,
gpxlogger
would quit in this case\&.
.RE
.PP
\fB\-V\fR, \fB\-\-version\fR
.RS 4
Dump the package version and exit\&.
.RE
.SH "SEE ALSO"
.PP
\fBgpsd\fR(8),
\fBgps\fR(1)
\fBgpspipe\fR(1)
.SH "AUTHORS"
.PP
Amaury Jacquot
<sxpert@sxpert\&.org>
& Petter Reinholdtsen
<pere@hungry\&.com>
& Chris Kuethe
<chris\&.kuethe@gmail\&.com>


debian/changelog
gpsd (3.22-2) unstable; urgency=medium

  [ Bernd Zeimetz ]
  * [66691096] Update Apparmor profile from upstream git HEAD.
    Thanks to Christian Ehrhardt (Closes: #976152) (LP: #1894330)

  [ Christian Ehrhardt ]
  * [63f0cbf0] symbols: nanowait came back as (int, timespec*)
    Signed-off-by: Christian Ehrhardt <christian.ehrhardt@canonical.com>

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 13 Jan 2021 19:20:14 +0100

gpsd (3.22-1) unstable; urgency=medium

  [ Pino Toscano ]
  * [d78fef17] Limit libsystemd-dev & libudev-dev build-deps as linux-any

  [ Bernd Zeimetz ]
  * [5afe80fe] New upstream version 3.22
  * [d23d0669] Updating debian/control from debian/control.in
  * [fde91987] libgps-dev: mark as M-A: same.
    Thanks to Lisandro Damián Nicanor Pérez Meyer (Closes: #912580)

 -- Bernd Zeimetz <bzed@debian.org>  Sat, 09 Jan 2021 22:42:25 +0100

gpsd (3.21.99~rc1-2) unstable; urgency=medium

  * Uploading to unstable
  * [ac4c2cbe] add automatically generated rdep builder
  * [3420d64b] Revert "Updating gbp.conf for experimental"
    This reverts commit c524723236f6334db4605752172412707e7f2cb3.

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 06 Jan 2021 01:36:36 +0100

gpsd (3.21.99~rc1-1) experimental; urgency=medium


  * [7936993f] Fixing git URLs
  * [3296cd58] removing extra linebreak
  * [8faaf157] Updating debian/control from debian/control.in
  * [c5247232] Updating gbp.conf for experimental
  * [4e2f5ac1] New upstream version 3.21.1~dev
  * [e01b4392] refreshing patches
  * [82d5b474] Addint matplotlib as dependency
  * [b905b4a6] snapshot changelog
  * [d415ffbd] remove gpsd build dir
  * [2a2d7e26] move dh_clean in order
  * [792171ef] Remove debian/source/options
  * [45c11a24] also remove buildtmp symlink
  * [b72ada42] add asciidoctor as build-dep
  * [704c85c4] Updating debian/control from debian/control.in
  * [8fa23322] remove gpsd-dbg package
  * [bf066735] Updating debian/control from debian/control.in
  * [18cf7d90] things happen in Sconscript, not sconstruct
  * [13d41b86] copy files from build dir
  * [448b12a0] Really ignore xgps deps check
  * [6b612ecf] add all the python stuff as build-dep
    tools are executed at build time
  * [61c389fe] Updating debian/control from debian/control.in
  * [04d323b3] python modules are pure python now
    no need to build for each python release
  * [c428275b] Build static library first, then the shared version
  * [a60bae70] add libsystemd-dev as build-dep
  * [5d5b5749] Updating debian/control from debian/control.in
  * [0ef4040b] add libudev-dev as build-dep
  * [3c09ab29] Updating debian/control from debian/control.in
  * [4cc239f6] tidy build-dir after static build
  * [a3991765] Fix filename of static library
  * [943cdb1e] No python c modules anymore
  * [7a0a4c1a] more fixes for SConstruct -> SConscript migration
  * [cf9802b0] scons fails to check xsltproc.
    force it
  * [520d0fb6] Don't call scons for clean.
    Just remove the builddir.
  * [3d6ad4ee] Remove upstream installed documentation
  * [01a55541] Install new gpsd-client tools.
  * [8ea6798e] install libgpsdpacket into python3-gps
  * [24703969] no python debug build anymore
  * [525cd475] Fix typo
  * [cf8b732e] install systemd service files
  * [1a5ac994] Update gpsd-clients.examples
  * [6ac6556e] install examples with dh_install
    dh_installexamples doesn't support #!/usr/bin/dh-exec....
  * [8bcc5152] Fix udev source location
  * [b317fad6] Updating symbols file
  * [90e3c370] Updating libgps symbols file.
    Also sorting it.
  * [112fdad9] Use scons systemd_install
  * [c2de2379] Add missing "" for c++ symbols
  * [88af2338] fix symbols
  * [5d4c7e4f] mark one more symbol as missing
  * [1814e3c1] New upstream version 3.21.99~rc1
  * [58d93d7d] Remove patch that is not needed anymore
  * [92242a31] snapshot changelog
  * [155be863] Updating debian/control from debian/control.in
  * [f2668dc0] try to build reverse-dep automatically
  * [44310e1a] gitlab ci fixes
  * [74f7bf91] actually enable job
  * [84eed5c3] assume yes for apt
  * [d557db05] remove old X-Python settings
  * [19770775] fix python3 shebangs
  * [dde6ae6a] add some more -y for apt
  * [f142e929] remove python2 shebang replacements

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 04 Jan 2021 17:05:16 +0100

gpsd (3.20-12) unstable; urgency=medium

  [ Bernd Zeimetz ]
  * [b7a30c42] Disable reprotest on CI for now.
    Until somebody comes up with a fix.

  [ Christian Ehrhardt ]
  * [fccbcc8f] fix json missing quotes on device add (LP: #1879272)
    Signed-off-by: Christian Ehrhardt <christian.ehrhardt@canonical.com>

 -- Bernd Zeimetz <bzed@debian.org>  Wed, 20 May 2020 09:04:54 +0200

gpsd (3.20-11) unstable; urgency=medium

  [ Christian Ehrhardt ]
  * [232c8d73] d/rules: fix ubxtool to use python3 in the gpsd package (LP: #1878158)
    Signed-off-by: Christian Ehrhardt <christian.ehrhardt@canonical.com>

 -- Bernd Zeimetz <bzed@debian.org>  Tue, 12 May 2020 13:49:15 +0200

gpsd (3.20-10) unstable; urgency=medium

  [ Christian Ehrhardt ]
  * [7670b61e] exchange gpsd-client/tools content
  * [fb81a96a] examples also need to be in -clients instead of -tools then
  * [7f8d06a8] d/control[.in]: moving gpsctl still breaks gpsd-clients
    (<< 3.20-9) for upgraders never seeing the version in experimental

  [ Bernd Zeimetz ]
  * [b448fb08] Merge branch 'exchange-clients-and-tools' into 'master'
    exchange gpsd-client/tools content
    See merge request debian-gps-team/pkg-gpsd!6

 -- Bernd Zeimetz <bzed@debian.org>  Mon, 11 May 2020 14:40:08 +0200

gpsd (3.20-9) experimental; urgency=medium

  [ Christian Ehrhardt ]
  * [6a21e6bd] d/usr.sbin.gpsd: improve apparmor rules for PPS
               usage (LP: #1872175 LP: #1872178)
  * [c8b703b4] d/gpsd.default: add USBAUTO option (LP: #1873415)
  * [2fcf9754] split more uncommon tools to gpsd-tools (LP: #1872189)
  * [4a1454fb] d/control[.in]: have gpsd-tools depend on gpsd-clients
               as it extends on cgps

gpxlogger.c
/* $Id: gpxlogger.c 4542 2007-12-12 20:03:05Z ckuethe $ */
#include <sys/types.h>
#include <string.h>
#include <stdlib.h>
#include <syslog.h>
#include <math.h>
#include <time.h>
#include <signal.h>

#include <glib.h>
#include <dbus/dbus.h>
#include <dbus/dbus-glib-lowlevel.h>
#include <dbus/dbus-glib.h>

#include <glib/gprintf.h>
#include "gpsd_config.h"
#include "gps.h"

DBusConnection* connection;

static char *author = "Amaury Jacquot";
static char *copyright = "BSD or GPL v 2.0";

static int intrack = 0;
static int first = 1;
static time_t tracklimit = 5; /* seconds */

static struct {
	time_t	old_time;
	time_t	time;
	gint32	mode;
	gdouble	ept;
	gdouble	latitude;
	gdouble	longitude;
	gdouble eph;
	gdouble altitude;
	gdouble epv;
	gdouble	track;
	gdouble epd;
	gdouble	speed;
	gdouble	eps;
	gdouble	climb;
	gdouble epc;
	//gdouble separation;
} gpsfix;


static void print_gpx_trk_start (void) {
	fprintf (stdout, " <trk>\n");
	fprintf (stdout, "  <trkseg>\n");
	fflush (stdout);
}

static void print_gpx_trk_end (void) {
	fprintf (stdout, "  </trkseg>\n");
	fprintf (stdout, " </trk>\n");
	fflush (stdout);
}

static DBusHandlerResult handle_gps_fix (DBusMessage* message) {
	DBusError	error;
	double		temp_time;

	dbus_error_init (&error);

	dbus_message_get_args (message,
			       &error,
			       DBUS_TYPE_DOUBLE, &temp_time,
			       DBUS_TYPE_INT32,	 &gpsfix.mode,
			       DBUS_TYPE_DOUBLE, &gpsfix.ept,
			       DBUS_TYPE_DOUBLE, &gpsfix.latitude,
			       DBUS_TYPE_DOUBLE, &gpsfix.longitude,
			       DBUS_TYPE_DOUBLE, &gpsfix.eph,
			       DBUS_TYPE_DOUBLE, &gpsfix.altitude,
			       DBUS_TYPE_DOUBLE, &gpsfix.epv,
			       DBUS_TYPE_DOUBLE, &gpsfix.track,
			       DBUS_TYPE_DOUBLE, &gpsfix.epd,
			       DBUS_TYPE_DOUBLE, &gpsfix.speed,
			       DBUS_TYPE_DOUBLE, &gpsfix.eps,
			       DBUS_TYPE_DOUBLE, &gpsfix.climb,
			       DBUS_TYPE_DOUBLE, &gpsfix.epc,
			       DBUS_TYPE_INVALID);
	gpsfix.time = floor(temp_time);
	
	/* 
	 * we have a fix there - log the point
	 */
	if ((gpsfix.time!=gpsfix.old_time)&&gpsfix.mode>1) {
		struct tm 	time;

		/* Make new track if the jump in time is above
		 * tracklimit.  Handle jumps both forward and
		 * backwards in time.  The clock sometimes jump
		 * backward when gpsd is submitting junk on the
		 * dbus. */
		if (fabs(gpsfix.time - gpsfix.old_time) > tracklimit && !first) {
			print_gpx_trk_end();
			intrack = 0;
		}

		if (!intrack) {
			print_gpx_trk_start();
			intrack = 1;
			if (first)
			    first = 0;
		}
		
		gpsfix.old_time = gpsfix.time;
		fprintf (stdout, "   <trkpt lat=\"%f\" lon=\"%f\">\n", gpsfix.latitude, gpsfix.longitude);
		fprintf (stdout, "    <ele>%f</ele>\n", gpsfix.altitude);
		gmtime_r (&(gpsfix.time), &time);
		fprintf (stdout, "    <time>%04d-%02d-%02dT%02d:%02d:%02dZ</time>\n",
				time.tm_year+1900, time.tm_mon+1, time.tm_mday,
				time.tm_hour, time.tm_min, time.tm_sec);
		if (gpsfix.mode==1)
			fprintf (stdout, "    <fix>none</fix>\n");
		else
			fprintf (stdout, "    <fix>%dd</fix>\n", gpsfix.mode);
		fprintf (stdout, "   </trkpt>\n");
		fflush (stdout);
	}
	return DBUS_HANDLER_RESULT_HANDLED;
}

static void print_gpx_header (void) {
	fprintf (stdout, "<?xml version=\"1.0\" encoding=\"utf-8\"?>\n");
	fprintf (stdout, "<gpx version=\"1.1\" creator=\"navsys logger\"\n");
	fprintf (stdout, "        xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\"\n");
	fprintf (stdout, "        xmlns=\"http://www.topografix.com/GPX/1.1\"\n");
	fprintf (stdout, "        xsi:schemaLocation=\"http://www.topografix.com/GPS/1/1\n");
	fprintf (stdout, "        http://www.topografix.com/GPX/1/1/gpx.xsd\">\n");
	fprintf (stdout, " <metadata>\n");
	fprintf (stdout, "  <name>NavSys GPS logger dump</name>\n");
	fprintf (stdout, "  <author>%s</author>\n", author);
	fprintf (stdout, "  <copyright>%s</copyright>\n", copyright);
	fprintf (stdout, " </metadata>\n");
	fflush (stdout);
}

static void print_gpx_footer (void) {
	if (intrack)
		print_gpx_trk_end();
	fprintf (stdout, "</gpx>\n");
	fclose (stdout);
}

static void quit_handler (int signum) {
	syslog (LOG_INFO, "exiting, signal %d received", signum);
	print_gpx_footer ();
	exit (0);
}

/*
 * Message dispatching function
 *
 */
static DBusHandlerResult signal_handler (
		DBusConnection* connection, DBusMessage* message) {
	/* dummy, need to use the variable for some reason */
	connection = NULL;
	
	if (dbus_message_is_signal (message, "org.gpsd", "fix")) 
		return handle_gps_fix (message);
	/*
	 * ignore all other messages
	 */
	
	return DBUS_HANDLER_RESULT_NOT_YET_HANDLED;
}

int main (int argc, char** argv) {
	GMainLoop* mainloop;
	DBusError error;

	/* initializes the gpsfix data structure */
	gps_clear_fix(&gpsfix);

	/* catch all interesting signals */
	signal (SIGTERM, quit_handler);
	signal (SIGQUIT, quit_handler);
	signal (SIGINT, quit_handler);

	//openlog ("gpxlogger", LOG_PID | LOG_NDELAY , LOG_DAEMON);
	//syslog (LOG_INFO, "---------- STARTED ----------");
	
	print_gpx_header ();
	
	mainloop = g_main_loop_new (NULL, FALSE);

	dbus_error_init (&error);
	connection = dbus_bus_get (DBUS_BUS_SYSTEM, &error);
	if (dbus_error_is_set (&error)) {
		syslog (LOG_CRIT, "%s: %s", error.name, error.message);
		return 3;
	}
	
	dbus_bus_add_match (connection, "type='signal'", &error);
	if (dbus_error_is_set (&error)) {
		syslog (LOG_CRIT, "unable to add match for signals %s: %s", error.name, error.message);
		return 4;
	}

	if (!dbus_connection_add_filter (connection, (DBusHandleMessageFunction)signal_handler, NULL, NULL)) {
		syslog (LOG_CRIT, "unable to register filter with the connection");
		return 5;
	}
	
	dbus_connection_setup_with_g_main (connection, NULL);

	g_main_loop_run (mainloop);
	return 0;
}
